subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -0,0 +1,2482 @@
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_LWM2M_LEVEL
+#include <logging/sys_log.h>
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+#include <ctype.h>
+#include <errno.h>
+#include <init.h>
+#include <misc/printk.h>
+#include <net/net_ip.h>
+#include <net/net_pkt.h>
+#include <net/zoap.h>
+#include <net/lwm2m.h>
+
+#include "lwm2m_object.h"
+#include "lwm2m_engine.h"
+#include "lwm2m_rw_plain_text.h"
+#include "lwm2m_rw_oma_tlv.h"
+#ifdef CONFIG_LWM2M_RW_JSON_SUPPORT
+#include "lwm2m_rw_json.h"
+#endif
+#ifdef CONFIG_LWM2M_RD_CLIENT_SUPPORT
+#include "lwm2m_rd_client.h"
+#endif
+
+#define ENGINE_UPDATE_INTERVAL 500
+
+/* LWM2M / CoAP Content-Formats */
+#define LWM2M_FORMAT_PLAIN_TEXT		0
+#define LWM2M_FORMAT_APP_LINK_FORMAT	40
+#define LWM2M_FORMAT_APP_OCTET_STREAM	42
+#define LWM2M_FORMAT_APP_EXI		47
+#define LWM2M_FORMAT_APP_JSON		50
+#define LWM2M_FORMAT_OMA_PLAIN_TEXT	1541
+#define LWM2M_FORMAT_OMA_OLD_TLV	1542
+#define LWM2M_FORMAT_OMA_OLD_JSON	1543
+#define LWM2M_FORMAT_OMA_OLD_OPAQUE	1544
+#define LWM2M_FORMAT_OMA_TLV		11542
+#define LWM2M_FORMAT_OMA_JSON		11543
+
+#define DISCOVER_PREFACE	"</.well-known/core>;ct=40"
+
+struct observe_node {
+	sys_snode_t node;
+	struct net_context *net_ctx;
+	struct sockaddr addr;
+	struct lwm2m_obj_path path;
+	u8_t  token[8];
+	s64_t event_timestamp;
+	s64_t last_timestamp;
+	u32_t min_period_sec;
+	u32_t max_period_sec;
+	u32_t counter;
+	bool used;
+	u8_t  tkl;
+};
+
+#define NUM_PENDINGS	CONFIG_LWM2M_ENGINE_MAX_PENDING
+#define NUM_REPLIES	CONFIG_LWM2M_ENGINE_MAX_REPLIES
+
+struct zoap_pending pendings[NUM_PENDINGS];
+struct zoap_reply replies[NUM_REPLIES];
+struct k_delayed_work retransmit_work;
+
+static struct observe_node observe_node_data[CONFIG_LWM2M_ENGINE_MAX_OBSERVER];
+
+static sys_slist_t engine_obj_list;
+static sys_slist_t engine_obj_inst_list;
+static sys_slist_t engine_observer_list;
+
+/* periodic / notify / observe handling stack */
+static K_THREAD_STACK_DEFINE(engine_thread_stack,
+			     CONFIG_LWM2M_ENGINE_STACK_SIZE);
+static struct k_thread engine_thread_data;
+
+/* for debugging: to print IP addresses */
+char *sprint_ip_addr(const struct sockaddr *addr)
+{
+	static char buf[NET_IPV6_ADDR_LEN];
+
+#if defined(CONFIG_NET_IPV6)
+	if (addr->family == AF_INET6) {
+		return net_addr_ntop(AF_INET6, &net_sin6(addr)->sin6_addr,
+				     buf, sizeof(buf));
+	}
+#endif
+#if defined(CONFIG_NET_IPV4)
+	if (addr->family == AF_INET) {
+		return net_addr_ntop(AF_INET, &net_sin(addr)->sin_addr,
+				     buf, sizeof(buf));
+	}
+#endif
+
+	SYS_LOG_ERR("Unknown IP address family:%d", addr->family);
+	return NULL;
+}
+
+char *sprint_token(const u8_t *token, u8_t tkl)
+{
+	int i;
+	static char buf[32];
+	int pos = 0;
+
+	for (i = 0; i < tkl; i++) {
+		pos += snprintf(&buf[pos], 31 - pos, "%x", token[i]);
+	}
+	buf[pos] = '\0';
+	return buf;
+}
+
+/* observer functions */
+
+int engine_notify_observer(u16_t obj_id, u16_t obj_inst_id, u16_t res_id)
```

Comment:
Yes, I suppose this could collide (in theory).  I'll add a lwm2m_* prefix to these shared functions, even though they are only used by the subsys/net/lib/lwm2m files.